Removes unused spec related to legacy checkout

filipefurtad0 opened the pull request:
Closes https://github.com/openfoodfoundation/openfoodnetwork/issues/12439.

Removes a legacy spec (`./spec/requests/checkout/failed_checkout_spec.rb`) and updates an existing spec (`./spec/controllers/checkout_controller_spec.rb`).

#### What? Why?

Contributes to #10958 (first bullet point)

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

From previous specs, attempting to checkout with a payment with `source = nil`, related a payment on legacy checkout, with a Stripe SCA payment method, triggered a error 400, preventing checkout completion.

This seems not to be the case on the current checkout.

The legacy spec has been removed and the existing controller spec updated to reflect this.

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- green build.

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

rioug commented on spec/requests/checkout/failed_checkout_spec.rb:
```diff
@@ -1,93 +0,0 @@
-      shipping_method.save!
-    end
-
-    it "clears shipments and payments before rendering the checkout" do
-      put update_checkout_path, params:, as: :json
-
-      # Checking out a bogus Stripe Gateway without a source fails at :payment
-      # Shipments and payments should then be cleared before rendering checkout
-      expect(response.status).to be 400
-      expect(flash[:error])
-        .to eq 'Payment could not be processed, please check the details you entered'
-      order.reload
-      expect(order.shipments.count).to be 0
-      expect(order.payments.count).to be 0
-      expect(order.adjustment_total).to eq 0
```
Do we need to cover this scenario for the new checkout ?

filipefurtad0 replied:
> Do we need to cover this scenario for the new checkout ?

Thanks, indeed this is the question I wanted to raise. I don't know if we reach this state on the current checkout - a (Stripe) payment entity without a source. If we do, then I think we should expect a status 400, and not 302. If we don't, then I guess there is no point having this new test.

For https://github.com/openfoodfoundation/openfoodnetwork/issues/10958 I'm trying to find a correspondence between old and new specs. I might run into such situations more often.

filipefurtad0 replied:
Ok, I've added a commit to undo changes to this file, as the test is perhaps nit particularly useful.

rioug replied:
> Thanks, indeed this is the question I wanted to raise. I don't know if we reach this state on the current checkout - a (Stripe) payment entity without a source. If we do, then I think we should expect a status 400, and not 302. If we don't, then I guess there is no point having this new test.

I don't know enough about payment and source to know if it's plausible scenario, happy to have a look if needed. I am wondering if the new checkout does not return a 400 in this scenario but display an error to the user instead.

dacook replied:
It sounds like we should try to replicate it with a feature spec. If not replicable, then I guess we can move on.

filipefurtad0 replied:
It's tough to replicate Stripe errors on feature specs - we'd need a proxy to intercept HTTP requests to the real Stripe... I've tried this with puffing billy but did not succeed.

This [error just came in on bugsnag](https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/66329690f4b6380007e8a4f8?event_id=6632969000e6f40fb4000000), it looks somewhat related (touches the checkout controller and the payment source). So maybe this scenario is relevant after all. I'll have a more thorough look tomorrow.

dacook replied:
Good pickup on the error!

I noticed that some data was blank in the request. Perhaps (hopefully) the credit card details were scrubbed before logging, but I wonder about `gateway_payment_profile_id`.

```
{
  "_method": "put",
...
  "order": {
    "payments_attributes": [
      {
        "source_attributes": {
...
          "month": "",
          "year": "",
          "cc_type": "",
          "last_digits": "",
          "gateway_payment_profile_id": ""
        },
        "payment_method_id": "201053"
      }
    ]
  },
  "commit": "Next - Order summary"
}
```

It gets set [a few different ways](https://github.com/search?q=repo%3Aopenfoodfoundation%2Fopenfoodnetwork+gateway_payment_profile_id+NOT+path%3A%2F%5Espec%5C%2F%2F+NOT+path%3Adb&type=code) in the codebase, but perhaps it's getting lost somehow.

filipefurtad0 replied:
Interesting. The request points to `"payment_method_id": "201053"` which - by looking at the production DB using Metabase - leads to a `Spree::PaymentMethod::Check` payment method. It seems we're getting this error across production servers (an error 500 on checkout :eyes: ). Below, also on FR-prod:

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/acb66e01-f3ba-4abd-8d7d-c3bbdf118d9f)

One our codebase, we have [the line](https://github.com/openfoodfoundation/openfoodnetwork/blob/2e004a026b8be846d12075382b201ab56b0df8c7/app/models/spree/payment_method.rb#L83):
`    # nil means the payment method doesn't require a source e.g. check`

I wonder if this happens due to a sort of race condition in which initially a Stripe payment method is selected for checkout but, during this time, disabled but the admin... or so. Since it relates to checkout, I think it might deserve an issue.

filipefurtad0 replied:
Ok, I've tried to reproduce the bugsnag error with a spec:
- with a cash payment method, we can trigger the error message, but somehow we get a status 200 (I was expecting 500')
- with Stripe payment method, we can see the data is incomplete, and just get the expected 422, with no error message

I'm not sure how to proceed on this one; do we expect payment_source_class to be required for payment methods with `type: "Spree::PaymentMethod::Check"`?

dacook replied:
Good find Filipe! I think the comment is right, we should just make it `nil` in this case. I've updated the code and the spec to show the error is now avoided. How does that look?

rioug reviewed (approved):
David got there before I did :). I think it looks good now.

dacook commented:
Sorry I forgot to press submit on my comments 😀

filipefurtad0 commented:
> Good find Filipe! I think the comment is right, we should just make it nil in this case. I've updated the code and the spec to show the error is now avoided. How does that look?

Awesome, @dacook . I think this can have an impact on checkout, so:
- I've opened an issue summarizing the issue (not too much info, but still helpful to have context I think)
- changed this issue to type `user facing`

We do have system specs for checkout with cash/check. Do you think it's worth having a round of manual testing?

dacook commented:
Great job Filipe 👍 

> We do have system specs for checkout with cash/check. Do you think it's worth having a round of manual testing?

I'm not sure, but given that we don't have a system spec for this case and don't know how to replicate it, I think it would be worth a manual test.  Just to see if we can replicate or learn anything more about it. Would you be willing to do that?

dacook commented:
I just realised we only have one review for my code change. I'll request Maikel's review before proceeding.

mkllnk commented on spec/controllers/checkout_controller_spec.rb:
```diff
@@ -308,6 +308,62 @@
         end
       end
 
+      context "with no payment source" do
```
FYI: you referenced a Bugsnag issue but Bugsnags retention is quite limited and it a month or so the link won't be valid anymore.

mkllnk reviewed (approved):
This makes sense.

mkllnk commented:
I think that we should check the checkout on staging though, with a cash payment method.

drummer83 commented:
Hi @filipefurtad0,
Thanks for working on this one! :muscle: 

I have staged the PR and tested the following:
- Build is green. :heavy_check_mark: 
- Check out with cash payment method. :heavy_check_mark: 
- Check Bugsnag for errors. :heavy_check_mark: 
- Check browser console for errors. :question: 

We do see some errors in the console. I am just now staging master to check if we have it there too. Probably it's unrelated. We shall see.

Here is the console output:

```
Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('https://js.stripe.com') does not match the recipient window's origin ('https://staging.openfoodnetwork.org.au'). summary
Uncaught Error: frame window is not ready
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    retryTimeout https://js.stripe.com/v3/:1
    __trace__ https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    setTimeout handler*Ye/load/r/< https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    value https://js.stripe.com/v3/:1
    promise callback*value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    b https://js.stripe.com/v3/:1
    _handleMessage https://js.stripe.com/v3/:1
    l https://js.stripe.com/v3/:1
    _setupPostMessage https://js.stripe.com/v3/:1
    __trace__ https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    r https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    addEventListener https://js.stripe.com/v3/:1
    _setupPostMessage https://js.stripe.com/v3/:1
    e https://js.stripe.com/v3/:1
    Ka https://js.stripe.com/v3/:1
    Ya https://js.stripe.com/v3/:1
    Xa https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    e https://js.stripe.com/v3/:1
    Wc https://js.stripe.com/v3/:1
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    mutationObserver https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    n https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    n https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
v3:1:111066
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    retryTimeout https://js.stripe.com/v3/:1
    __trace__ https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    (Async: setTimeout handler)
    r https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    value https://js.stripe.com/v3/:1
    (Async: promise callback)
    value https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    b https://js.stripe.com/v3/:1
    _handleMessage https://js.stripe.com/v3/:1
    l https://js.stripe.com/v3/:1
    _setupPostMessage https://js.stripe.com/v3/:1
    __trace__ https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    (Async: EventListener.handleEvent)
    r https://d2wy8f7a9ursnm.cloudfront.net/v7/bugsnag.min.js:1
    addEventListener https://js.stripe.com/v3/:1
    _setupPostMessage https://js.stripe.com/v3/:1
    e https://js.stripe.com/v3/:1
    Ka https://js.stripe.com/v3/:1
    Ya https://js.stripe.com/v3/:1
    Xa https://js.stripe.com/v3/:1
    value https://js.stripe.com/v3/:1
    e https://js.stripe.com/v3/:1
    Wc https://js.stripe.com/v3/:1
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    forEach self-hosted:203
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    mutationObserver https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    (Async: MutationCallback)
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    e https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    value https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    n https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    n https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2
    <anonymous> https://staging.openfoodnetwork.org.au/packs/js/application-d68d15335f2ff6bf3cd4.js:2

```

I have to stop here and check master later.

Thanks again!

drummer83 commented:
Sorry it took so long, @filipefurtad0!

I can confirm that the errors are in master as well. Unfortunately there is a conflict now. Could you have a look at that please? Let me know if I should re-test afterwards.

Thanks and sorry again! 🙏

filipefurtad0 commented:
> I can confirm that the errors are in master as well. Unfortunately there is a conflict now. Could you have a look at that please? Let me know if I should re-test afterwards.

Hey @drummer83 ,

All good - thank you for testing.
The conflict arose due to a [change on a file](https://github.com/openfoodfoundation/openfoodnetwork/commit/6317fe1b717b0dd569f67a514e7dec88d56980ed#diff-1b2f151d1a06126279b58f072a1546ce57fe263cf74a5642bcf05615496784fe) this PR is removing; I've confirmed the file removal during rebase which fixes the conflict.

Everything else is left untouched. I would say it's ok to merge straight away.

dacook commented:
Merged, thanks all!

dacook commented:
Although this change touches a small amount of checkout code, I don't think it changed any behaviour, so I'm changing the tag to `technical changes only`.
